### DISCUSSION TITLE
returning no effect in an effect mapping

### DISCUSSION THREAD

**Discussion by yspreen** [pointfreeco/swift-composable-architecture#1068] [Q&A] [answered]
we have this pattern where we frequently create a `Effect<Result<Response, Error>, Never>` inside our reducers, which is then getting mapped to a new action. but some of those results don't result in new actions. this leads to the following code:

```
case .someAction:
        return environment.dataClient.apiCall().map { result in
            if result.someCondition {
              return .newAction
            }
            return .none
        }
```

we can't just return nil because that's an invalid mapping. this means we have to add a `none` case to all of our action enums. is there a better way for that? like a conditional fire and forget or something?

**Comment by importRyan** [comment]
Undoubtedly there are other practices, but our pattern is to have an apiCallFinished action that accepts the `Result<Response, Error>` to limit the scope of responsibilities. The reaction to someCondition is handled in that action.

```swift
    return environment.dataClient
      .apiCall()
      .receive(on: environment.mainQueue)
      .catchToEffect(Action.apiCallFinished)
      .cancellable(id: APICallId(), cancelInFlight: true)

  case let .apiCallFinished(.success(response)):
    // if some condition
    return .none

  case let . apiCallFinished(.failure(error)):
    // handle error
    return .none
```

**Comment by stephencelis** [comment, marked as answer]
> we can't just return nil because that's an invalid mapping.

Instead of `map`, you can use Combine's `compactMap`, which lets you filter out `nil` outputs in the closure. Just make sure you call `.eraseToEffect()` on the publisher it returns to get things back in an `Effect`.

**Reply by yspreen** [reply]
genius!

**Comment by lukeredpath** [comment]
I’d recommend the approach in the post of above and handle this logic in your reducer, but you can also use flatMap and conditionally return an Effect(value:) or Effect.none.
